WordPress/gutenberg
# Scripts: Update webpack and related dependencies to the latest version

## What?


Let's bring webpack and related dependencies to the latest version and apply them to WordPress core around 6.4 Beta 1:

- `webpack` updated to the latest version
- `browserslist` and `caniuse-lite` updated to the latest version
- `react-refresh` updated to the latest version

In addition, it tries to fix https://github.com/WordPress/gutenberg/issues/53552 caused by the outdated version of `webpack-cli`:

> When ncu was executed, @wordpress/scripts was displayed as ^26.9.0 → ^26.10.0, so ncu -u npm install was executed in sequence.
> As a result, several npm modules were shown as deprecated as follows:
> 
> npm WARN deprecated debuglog@1.0.1: Package no longer supported. Contact Support at https://www.npmjs.com/support for more info.
> npm WARN deprecated readdir-scoped-modules@1.1.0: This functionality has been moved to @npmcli/fs
> npm WARN deprecated @npmcli/move-file@1.1.2: This functionality has been moved to @npmcli/fs
> npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs

## Why?


> These issues has been resolved in https://github.com/webpack/webpack-cli/issues/3891, https://github.com/webpack/webpack-cli/issues/3892, and https://github.com/webpack/webpack-cli/issues/3893 reported in webpack-cli, so it is appropriate to replace it with the latest version.

## How?


It brings `webpack-cli` to the latest version. There were some breaking changes in [v5.0.0](https://github.com/webpack/webpack-cli/releases/tag/webpack-cli%405.0.0), but they don't seem to have impact like:
-  remove the `--hot` option in favor of directly using the `HotModuleReplacement` plugin (only for build command, for serve it will work). I don't see the listed deprecated packages in the lock file anymore.

There was a recent version bump enforced for the `webpack` in the repository with https://github.com/WordPress/gutenberg/pull/49075, so here we follow-up with another version bump that also gets applied to `@wordpress/scripts`.

## Testing Instructions





1. Run `nvm use` to switch to the currently supported Node.js and npm versions.
2. Run `npm install` to install the latest version of dependencies.
3. Run `npm run build` and confirm the the build finishes correctly.
4. Run `npx wp-create-block example-static --no-wp-scripts` to confirm that Create Block still works.
5. Run `cd example-statice` to enter the newly created project.
6. Run `../node_modules/.bin/wp-scripts start --hot` to confirm that you can develop blocks with the dev server.
7. Open the site at https://localhost:8888/wp-admin, go to the post editor and confirm that you can insert Example Static block.
8. Edit the JS code of the block in the editor. It must be something controlled by React, for example `edit` function. See that the block gets automatically updated in the post editor after saving changes locally.

## Screenshots or screencast 

I confirmed that React Fast Refresh still works with the updated `webpack-cli` and other dependencies:

<img width="970" alt="Screenshot 2023-09-20 at 11 56 08" src="https://github.com/WordPress/gutenberg/assets/699132/6fea4ab3-5b05-4bab-bbb4-d1807f3b5266">
